Use OkHttp

xerial opened the pull request:
- When td-client is used with jetty 9.4.4, it causes severe problems (especially when using Presto 0.187 + td-client at the same time). 
- Fixing various issues in #99 (jetty upgrade) seems difficult
- This PR also includes Guava, Jackson upgrades
- Targeting Java8 because Guava 21.0 or higher basically requires Java8.

miniway commented on pom.xml:
```diff
@@ -94,11 +94,25 @@
-      <version>${jetty.version}</version>
+      <groupId>com.squareup.okhttp3</groupId>
+      <artifactId>okhttp</artifactId>
+      <version>3.9.0</version>
```
3.8.1 is used at Presto 0.187 but guess it's compatible

xerial replied:
Yeah. Presto doesn't use OkHttp for server side code, so it might be OK.

miniway commented on src/main/java/com/treasuredata/client/TDClient.java:
```diff
@@ -406,7 +405,8 @@ public boolean existsTable(String databaseName, String tableName)
         catch (TDClientHttpException e) {
-            if (e.getStatusCode() == HttpStatus.NOT_FOUND_404) {
+            // NOT_FOUND error
+            if (e.getStatusCode() == 404) {
```
`java.net.HttpURLConnection.HTTP_NOT_FOUND` could be used

xerial replied:
I copied HttpStatus.java, but we can use HttpUrlConnection.xxx instead. Good suggestion.

xerial replied:
HttpStatus.java contains code name etc., which is useful for debugging and error message. Let me use this.

miniway replied:
Ah, you copied `HttpStatus.java`.

miniway commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -205,15 +188,7 @@ ObjectMapper getObjectMapper()
-                throw Throwables.propagate(e);
-            }
-        }
+        // No need to close OkHttp client
```
`com.facebook.presto.cli.QueryRunner` has some cleanup code for OkHttpClient. Don't we need it also? 
```java
    @Override
    public void close()
    {
        httpClient.dispatcher().executorService().shutdown();
        httpClient.connectionPool().evictAll();
    }
```

xerial replied:
I think we should pick this clean-up code. Thanks!

xerial commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -303,148 +288,188 @@ public String apply(Integer input)
 
-    public <ResponseType extends Response, Result> Result submitRequest(TDApiRequest apiRequest, Optional<String> apiKeyCache, Handler<ResponseType, Result> handler)
-            throws TDClientException
+    private class RequestContext
```
TODO: This should be a static class. We only need to pass some config parameters when initializing this class.

xerial commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -303,148 +288,188 @@ public String apply(Integer input)
-                        int code = re.getResponse().getStatus();
-                        throw handleHttpResponseError(apiRequest.getPath(), code, new byte[] {}, re.getResponse());
+            }
+            catch (Exception e) {
```
TODO: Make this handler pluggable

xerial commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -118,64 +109,55 @@ protected SimpleDateFormat initialValue()
+        this.httpClient = builder.build();
+        this.headers = ImmutableMultimap.copyOf(config.headers);
+
+//        if (config.requestBufferSize.isPresent()) {
```
It seems there is no configuration for the request/response buffer size. We should remove these configs

xerial commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -664,113 +703,84 @@ private ObjectReader getJsonReader(final JavaType type)
         }
+    }
 
+    public static class StringContentHandler
```
TODO: We only need one instance for this.

xerial commented on src/main/java/com/treasuredata/client/impl/ProxyAuthenticator.java:
```diff
@@ -20,44 +20,50 @@
     {
-        return proxyConfig.getUri();
-    }
+        if (response.request().header("Proxy-Authorization") != null) {
```
Without this line, OkHttp enters infinite-loop if a wrong password is used for 
 proxy servers: https://github.com/square/okhttp/wiki/Recipes#handling-authentication

miniway commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -303,148 +286,188 @@ public String apply(Integer input)
                 auth = apiKey.get();
             }
-            request.header(HttpHeader.AUTHORIZATION, auth);
+            request = request.header("Authorization", auth);
```
We can also use `com.google.common.net.HttpHeaders.AUTHORIZATION`

miniway commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -276,20 +260,19 @@ public String apply(Integer input)
+                        .url(requestUri)
+                        .header("User-Agent", getClientName())
+                        .header("Date", dateHeader);
 
```
We can use `com.google.common.net.HttpHeaders.*` constants

miniway commented on src/main/java/com/treasuredata/client/TDHttpClient.java:
```diff
@@ -513,7 +535,7 @@ private TDClientException handleHttpResponseError(String apiRequestPath, int cod
     static Date parseRetryAfter(long now, Response response)
     {
-        String retryAfter = response.getHeaders().get("Retry-After");
+        String retryAfter = response.header("Retry-After");
```
same here `HttpHeaders.RETRY_AFTER`

miniway commented on src/main/java/com/treasuredata/client/impl/ProxyAuthenticator.java:
```diff
@@ -20,44 +20,50 @@
-        return proxyConfig.getUri();
-    }
+        if (response.request().header("Proxy-Authorization") != null) {
+            throw new IOException(new TDClientHttpException(TDClientException.ErrorType.PROXY_AUTHENTICATION_FAILURE, "Proxy authentication failure", 407, null));
```
HttpHeaders.PROXY-AUTHORIZATION

miniway commented on src/test/java/com/treasuredata/client/TestServerFailures.java:
```diff
@@ -18,7 +18,6 @@
-import com.google.common.base.Throwables;
 import com.google.common.io.CharStreams;
 import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.server.Request;
```
https://gist.github.com/miniway/9a674feba93fd3b1bf5ee90b540a23fd

I think we can completely remove jetty dependency even in test cases

xerial commented:
@miniway (cc: @muga, @dmikurube, @k0kubun)
Added documentations and refactored codes to make clear the roles of classes (e.g. TDHttpClient handles http request/response, and TD API specific error handling are now in TDRequestErrorHandler) 

I'll merge this to master and release 0.8.0 based on this PR, and create 0.7.x branch to manage the old version for Java 7 users.
